public/app/features/plugins/pluginPreloader.ts

```diff
@@ -15,67 +13,28 @@ export type PluginPreloadResult = {
   addedLinkConfigs?: PluginExtensionAddedLinkConfig[];
 };
 
-export async function preloadPlugins(
-  apps: AppPluginConfig[] = [],
-  registries: PluginExtensionRegistries,
-  eventName = 'frontend_plugins_preload'
-) {
-  startMeasure(eventName);
-  const promises = apps.filter((config) => config.preload).map((config) => preload(config));
-  const preloadedPlugins = await Promise.all(promises);
+const preloadedAppPlugins = new Set<string>();
+const isNotYetPreloaded = ({ id }: AppPluginConfig) => !preloadedAppPlugins.has(id);
+const markAsPreloaded = (apps: AppPluginConfig[]) => apps.forEach(({ id }) => preloadedAppPlugins.add(id));
```

Comment:
Are these used to know what was already loaded? `markAsPreloaded` is called before the actual preloading? 

Also you could tap into the SystemJS registry to know if a plugin is already loaded. 🤔

Reply:
> Are these used to know what was already loaded? markAsPreloaded is called before the actual preloading so I'm not sure.

Basically my intention was to introduce a "cache" here, so I don't run the preloading logic multiple times for the same plugin. I called it before the actual preloading to fight racing conditions (if I do it once all the promises are resolved, then there is a time window while these apps can be preloaded again). On the other I should have probably remove them from the list of preloaded plugins if their preloading fails?

> Also you could tap into the SystemJS registry to know if a plugin is already loaded. 🤔

Yeah, I think this relates somewhat to my intention above, I was basically trying to guard against entering the preloading logic for the same plugin multiple times, while the SystemJS registry is only populated once all the preloading is finished, which means that we would have a time window where the same plugin can be preloaded multiple times - I am not sure if that makes sense?

Reply:
Ok, sounds like it's probably best we get this merged as is and we circle back to it to figure out removing from the list if preloading fails / tapping into systemjs to guard against running preload logic for same plugin multiple times. 👍